### DISCUSSION TITLE
Azure DevOps PR descriptions are truncated at 4000 characters

### DISCUSSION THREAD

**Discussion by ecraig12345** [renovatebot/renovate#16689] [Request Help]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

latest from npm, currently 32.120.0

### Please select which platform you are using if self-hosting.

Azure DevOps (dev.azure.com)

### If you're self-hosting Renovate, tell us what version of the platform you run.

latest cloud version

### Was this something which used to work for you, and then stopped?

I never saw this working

### Describe the bug

Azure DevOps has a [4000-character limit](https://developercommunity.visualstudio.com/t/raise-the-character-limit-for-pull-request-descrip/365708) for PR descriptions. 

Where we noticed this issue was in the onboarding PR: in a JS monorepo with a lot of packages, just the list of detected package.json files takes up most of the character limit, and the list of PRs to be created is truncated.

<details><summary>click to expand screenshots</summary>
<img width="755" alt="image" src="https://user-images.githubusercontent.com/5864305/180101094-cffc4d95-6e59-4e52-b7b5-f5577ff4a03f.png">
<img width="757" alt="image" src="https://user-images.githubusercontent.com/5864305/180101111-51f8a135-1308-4274-a0f1-61d9a2bbb647.png">
<img width="757" alt="image" src="https://user-images.githubusercontent.com/5864305/180101298-af3d59ab-4865-4c5a-9a99-c4363b58bc93.png">

</details>

This seems related but not identical to #2569, as this is an issue of (rather unreasonably low) hard limits on length specific to ADO. Also, the onboarding PR body doesn't appear to be configurable to reduce the length.

Sample repo with many packages: https://dev.azure.com/uifabric/fabricpublic/_git/renovate-public-demo

Onboarding PR for this repo (with truncated body, screenshots above): https://dev.azure.com/uifabric/fabricpublic/_git/renovate-public-demo/pullrequest/2933

### Relevant debug logs

<details><summary>Logs</summary>

[Full debug logs](https://dev.azure.com/uifabric/fabricpublic/_build/results?buildId=258523&view=logs&j=12f1170f-54f2-53f3-20dd-22fc7dff55f9&t=9c939e41-62c2-5605-5e05-fc3554afc9f5) (not sure which part is most relevant)

</details>

### Have you created a minimal reproduction repository?

I have linked to a minimal reproduction repository in the bug description

**Comment by viceice** [comment]
i don't see a bug here, that's just a DevOps limitation. i also don't think we can do anything to fix that

**Comment by ecraig12345** [comment]
It's a platform limitation that Renovate could handle better. Some ideas:
- Providing more intelligent truncation automatically for this platform
- Continuing the description in comments
- Making descriptions (including onboarding) more configurable

**Comment by viceice** [comment]
ok, it's still no bug.

**Comment by HonkingGoose** [comment]
Related issue:

- https://github.com/renovatebot/renovate/issues/16373

**Comment by ecraig12345** [comment]
> ok, it's still no bug.

Then change the issue type to feature request if you want? Based on [this](https://developercommunity.visualstudio.com/t/raise-the-character-limit-for-pull-request-descrip/365708) it seems highly unlikely that Azure DevOps will do anything about it, so unfortunately that means either Renovate+ADO users are stuck with a bad experience, or a workaround is needed in Renovate.

**Reply by rarkins** [reply]
I've converted the issue to a discussion. It takes more than just switching a label to convert a but report into a coherent feature request and would instead be confusing.

A new feature request issue can follow once a feasible enhancement idea has been found in this discussion.

Due to the very short character limitation, we will need to think about which sections should be stripped first in order to get under the limit.

**Reply by ecraig12345** [reply]
For the onboarding PR, my preference would probably be stripping or reducing the list of detected packages. The demo repo with ~60 packages is actually a lot smaller than some of the ones we'll be onboarding soon, and while the onboarding PR isn't absolutely essential, it's a nice way to give the repo owners (and the person setting it up) an idea of what's about to happen.

One way to reduce the list of detected packages would be listing the number of packages from each data source type, like "X npm packages" (to see if the number is roughly correct) and then they can look at the logs for more details. This could be done conditionally: maybe for ADO repos over a certain number of packages, or ADO repos if the default description is too long.

**Reply by ecraig12345** [reply]
For upgrade PRs, I still need to take a more detailed look at what truncation issues we're hitting and whether it can be addressed with manual customization of the existing `prBody*` options. (Though more intelligent auto-truncation would still be nice to have.)

Another idea would be putting the entire intended PR body in a debug log, so at least it would be visible somewhere.

**Reply by viceice** [reply]
DevOps is the only renovate platform which has a small body limit so no core renovate maintainer will have time to create a solution for that.

also the renovate codebase is already very complex and it's not easy to implement any smart truncation for a single platform because the platform code doesn't even know about the content, it just tries to truncate before sending the date to the platform.

**Reply by ecraig12345** [reply]
Okay, then tell me where to look in the codebase and I can probably contribute something (once a design is agreed upon)...? For context, I'm trying to get Renovate working against some large internal repos at Microsoft which unfortunately are stuck with ADO and all its limitations for the foreseeable future. The onboarding PR is less essential, but it's important to have the Renovate+ADO experience working as well as possible overall, and I'm willing to help make that happen.

**Reply by rarkins** [reply]
I think the key requirement would be the following?
- onboarding PR should not be accidentally *misleading*
- additional information can be found in debug logs

**Reply by ecraig12345** [reply]
Yes, that seems like a good summary (assuming you meant to write "should *not* be accidentally misleading").
